### PR TITLE
loss mask aware final layer applicaiton

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

save memory during training by loss mask selection for final output layer + softmax.

**Collection**: [NLP]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 1]
```diff
@@ -402,21 +402,24 @@ def inference_step(self, dataloader_iter, batch_idx, mode, dataloader_idx=0):
...
-
-        output = self.predict_step(batch, batch_idx, dataloader_idx)
-
-        inputs_text = [self.tokenizer.ids_to_text(c.tolist()) for c in batch['contexts']]
-        labels_text = [self.tokenizer.ids_to_text(a.tolist()) for a in batch['answers']]
-        preds_text = [
-            self.tokenizer.ids_to_text(t[l.item() :][: data_cfg.get('tokens_to_generate')])
-            for t, l in zip(output['token_ids'], batch['context_lengths'])
-        ]
+        if data_cfg.get("write_predictions_to_file", False) or data_cfg.metric.name != 'loss':
```
Comment (arendu) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:405]: This is a merge from another PR which is going to be merged soon.

File: nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py [hunk 2]
```diff
@@ -308,6 +315,12 @@ def forward(
                 sequence_parallel=self.sequence_parallel,
                 gradient_accumulation_fusion=self.config.gradient_accumulation_fusion,
             )
+            if loss_mask is not None:
```
Comment (stevehuang52) [inline nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:318]: `post_process_result` is a tuple of `(loss, logits)` when `encoder_input is not None`, which will break the line `res = torch.zeros_like(labels).type_as(post_process_result)`. Maybe consider this:

```python
if isinstance(post_process_result, tuple):
    loss, logits = post_process_result
 else:
    loss, logits = post_process_result, None
 
if loss_mask is not None:
    res = torch.zeros_like(labels).type_as(loss)
    res[loss_mask == 1] = loss
    if logits is not None:
        return res, logits
     else:
        return res
else:
    return post_process_result

Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:318]: good catch!

Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:318]: we should consider a CI test when encoder_input is not None.

**Review by stevehuang52 — APPROVED** [review]
I have tested the code locally and it worked. Thanks for the effort~!
